spec/index.bs

```diff
@@ -601,107 +601,122 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
-    {{GPUTextureUsage/SAMPLED|GPUTextureUsage.SAMPLED}}
-    don't change the contents of a resource.
-
-: <dfn dfn>Mutating usage</dfn>s
-:: Usages like {{GPUBufferUsage/STORAGE|GPUBufferUsage.STORAGE}}
-    do change the contents of a resource.
-
-Issue(gpuweb/gpuweb#296): Consider merging all read-only usages.
+A [=physical resource=] can be used on GPU in one of the following <dfn dfn>internal usage</dfn>s:
+<dl dfn-type=dfn dfn-for="internal usage">
+    : <dfn>input</dfn>
+    :: Buffer with input data for draw or dispatch calls. Preserves the contents.
+        Allowed by {{GPUBufferUsage/INDEX}}, {{GPUBufferUsage/VERTEX}}, or {{GPUBufferUsage/INDIRECT}}.
+    : <dfn>constant</dfn>
+    ::  Resource bindings that are constant from the shader point of view. Preserves the contents.
+        Allowed by {{GPUBufferUsage/UNIFORM}} or {{GPUTextureUsage/SAMPLED}}.
+    : <dfn>storage</dfn>
+    ::  Read-write storage resource binding.
+        Allowed by {{GPUBufferUsage/STORAGE}}.
+    : <dfn>storage-read</dfn>
+    ::  Read-only storage resource bindings. Preserves the contents.
+        Allowed by {{GPUBufferUsage/STORAGE}} or {{GPUTextureUsage/STORAGE}}.
+    : <dfn>storage-write</dfn>
+    ::  Write-only storage resource bindings.
+        Allowed by {{GPUTextureUsage/STORAGE}}.
+    : <dfn>attachment</dfn>
+    :: Texture used as an output attachment in a render pass.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+    : <dfn>attachment-read</dfn>
+    ::  Texture used as a read-only attachment in a render pass. Preserves the contents.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+</dl>
 
 Textures may consist of separate [=mipmap levels=] and [=array layers=],
 which can be used differently at any given time.
-Each such <dfn dfn>subresource</dfn> is uniquely identified by a
+Each such <dfn dfn>texture subresource</dfn> is uniquely identified by a
 [=texture=], [=mipmap level=], and
 (for {{GPUTextureDimension/2d}} textures only) [=array layer=].
 
-The **main usage rule** is that any [=subresource=]
-at any given time can only be in either:
-  - a combination of [=read-only usage=]s
-  - a single [=mutating usage=]
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
+
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage multiset</dfn>s, where each usage is compatible with the other:
+  - {[=internal usage/input=] | [=internal usage/constant=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage-write=]} - used one or more times
+  - [=internal usage/attachment=] - singular usage
+  - {[=internal usage/attachment-read=] | [=internal usage/storage-read=] | [=internal usage/constant=]} - in any combination
```

Comment:
As specified the allowed usage multisets should be infinite multisets, and I don't think that's clearly explained. E.g. the first one is {input &times; ∞, constant &times; ∞, storage-read &times; ∞} and the fourth one is {attachment &times; 1}.

Instead of defining the actual allowed multisets we can probably simply write them as expressions. Here's an incomplete suggestion:

```
<div algorithm=...>
    A [=usage multiset=] |U| is valid if it satisfies any of the following rules:
    - Each usage in |U| is [=internal usage/input=], [=internal usage/constant=], or [=internal usage/storage-read=].
    - Each usage in |U| is [=internal usage/storage=] or [=internal usage/storage-read=].
    - Each usage in |U| is [=internal usage/storage-write=].
    - Each usage in |U| is [=internal usage/attachment-read=], [=internal usage/storage-read=], or [=internal usage/constant=].
    - |U| contains exactly one element: [=internal usage/attachment=].
</div>
```

Alternative (but too wordy): could say something like "The set of usages in |U| is a subset of ..."

Reply:
I liked multisets and having a set of them for the mathematical purity. But I agree that changing the wording to be more to the point ends up with a cleaner spec. We can save the math purity for implementation details ;)
So I removed the mention of "multiset" now, operating on the lists, and used the definition you suggested here.